### PR TITLE
[feat] FIL 20 -- The Trash Edition

### PR DESCRIPTION
**Pull request by y-lohse**
First, I apologize to the person(s) who will review this PR, considering the number and nature of the changes. I'm going to annote and highlight the trickier changes, or the ones that may be improved.

This PR adds the Trash view. It's a whole new view, but it's also very similar to the existing Files view: you mainly browse files and folder. The differences are in the interactions, as you can only choose to restore some files, or delete them permanently (the latter is not part of this PR, only the UI bits).

### CODE REVIEW

File: src/actions/index.js [hunk 1]
```diff
@@ -51,6 +54,11 @@ export const downloadFileOffline = () => ({ type: DOWNLOAD_FILE_E_OFFLINE, alert
 
 export const openFolder = (folderId = ROOT_DIR_ID, isInitialFetch = false, router = null) => {
   return async dispatch => {
+    let routePrefix = '/files'
```
Comment (y-lohse) [inline src/actions/index.js]: `openFolder` is now used to navigate in `/files`and `/trash`, so having a hard-coded route prefix wasn't an option any more. The prefix could be supplied as an argument, but since we already have the router at hand, I think it's safe to extract the prefix from there.

File: src/actions/index.js [hunk 2]
```diff
@@ -193,8 +201,8 @@ export const deleteFileOrFolder = (id, isNew = false) => {
   }
 }
 
-export const trashFile = (id) => {
```
Comment (y-lohse) [inline src/actions/index.js]: Just some unrelated syntax changes here.

File: src/assets/icons/icon-trash-dark.svg [hunk 3]
```diff
@@ -0,0 +1,6 @@
+<svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" viewBox="0 0 16 16">
+  <g fill="#32363f" fill-rule="evenodd" transform="translate(-96 -32)">
```
Comment (y-lohse) [inline src/assets/icons/icon-trash-dark.svg]: @GoOz next on my wishlist: figuring out a way to change svg colors on the fly 😁

File: src/components/App.jsx [hunk 4]
```diff
@@ -1,30 +1,17 @@
...
 
-import { openFolder, alertClosed } from '../actions'
+import { alertClosed } from '../actions'
 
 import Alerter from 'cozy-ui/react/Alerter'
 import Sidebar from './Sidebar'
 import Topbar from './Topbar'
 
 class App extends Component {
-  componentWillMount () {
```
Comment (y-lohse) [inline src/components/App.jsx]: One of the major changes: the App component is no longer the one who initializes data load. Since it's used across the whole app and the data that needs to be loaded changes on each view, I feel like this is the wrong place to do it.
Now the data is loaded by the appropriate containers, eg. `Folder` and `Trash`. This is also consistent with what we do in cozy-photos AFAIK.

File: src/components/Nav.jsx [hunk 5]
```diff
@@ -9,6 +9,7 @@ import { Link, withRouter } from 'react-router'
 
 const Nav = ({ t, router }) => {
   let isBrowsingFiles = router.location.pathname.match(/^\/files/) !== null
+  let isBrowsingTrash = router.location.pathname.match(/^\/trash/) !== null
```
Comment (y-lohse) [inline src/components/Nav.jsx]: I feel like there should be a better, more global way to do this. As you can see later in this PR, I've even added a reducer that does it and used it in a couple of places. However, we haven't one for `isBrowsingFiles `, so maybe we should add that too? Then again, the router is a relatively good way of finding out what the user is viewing.

File: src/containers/Folder.jsx [hunk 6]
```diff
@@ -10,6 +10,18 @@ import FileList from '../components/FileList'
 const isDir = attrs => attrs.type === 'directory'
 
 class Folder extends Component {
+  componentWillMount () {
```
Comment (y-lohse) [inline src/containers/Folder.jsx]: So here we find the routing code that I removed from `App.jsx`.

File: src/reducers/index.js [hunk 7]
```diff
@@ -54,4 +54,6 @@ export const getActionableFile = ({ files, ui }) => getFileById(files, ui.action
 
 export const mustShowSelectionBar = state => state.ui.showSelectionBar || state.ui.selected.length !== 0
 
+export const isBrowsingTrash = state => state.folder && state.folder.path && state.folder.path.indexOf(TRASH_DIR_PATH) === 0
```
Comment (y-lohse) [inline src/reducers/index.js]: I'm not overly fond of th new `TRASH_DIR_PATH` constant, but I really can't see a better way to do this. The main problem is, we only ever know:

- what the current folder's id is
- what the parent's folder id is
- the id of that parent's dir

But if the current folder is 4 levels deep inside the trash, there's no way to find that information. Except in stringified path.

File: src/styles/breadcrumb.styl [hunk 8]
```diff
@@ -8,6 +8,8 @@
     color #95999d
     a
         cursor pointer
+    .separator
```
Comment (y-lohse) [inline src/styles/breadcrumb.styl]: Sort of unrelated to the PR, but the spacing in the breadcrumb was relying on literal spaces and was really inconsistent. Now it's handled with css!

File: src/assets/icons/icon-trash-dark.svg (hunk 3 above)
Comment (GoOz) [inline src/assets/icons/icon-trash-dark.svg]: Something that worked so far but I didn't get the chance to make it a rule is to fill the SVG with the `currentColor` keyword and then then change the `color` property of the element wrapping the SVG **BUT ONLY** when we're talking about inline SVGs, not in CSS background :P

File: src/styles/toolbar.styl [hunk 9]
```diff
@@ -6,6 +6,18 @@
     .coz-btn:last-child
         margin-right 0
 
+.fil-toolbar-trash
+    @extends $buttons
```
Comment (y-lohse) [inline src/styles/toolbar.styl]: @GoOz  and @m4dz : this one was an interesting problem. The related DOM is [here](https://github.com/y-lohse/cozy-files-v3/blob/FIL-20/src/containers/TrashToolbar.jsx#L15).

The problem I encountered was, the first button (`coz-btn--delete`) really needs the `danger-outline` class that is defined [here](https://github.com/cozy/cozy-ui/blob/v3/stylus/ui-components/button.styl#L118). But when I do that, I import the selector `[role=button], button`, and now [the other button](https://github.com/y-lohse/cozy-files-v3/blob/FIL-20/src/containers/TrashToolbar.jsx#L21) *also* inherits these styles -- because this selector is stronger than `coz-btn coz-btn--more`.

Is there a way to @extend only some parts of the placeholder? That would be quite handy in cases like this.

File: src/assets/icons/icon-trash-dark.svg (hunk 3 above)
Comment (y-lohse) [inline src/assets/icons/icon-trash-dark.svg]: Yes, this is what I used to do as well and it worked great. But as you said, it only works for inline SVG so I was only using that, whereas the current approach in cozy is to use background everywhere. Unless there's a reason to prefer backgrounds, I'd be in favor of switching to inline svg tbh.

File: src/styles/toolbar.styl (hunk 9 above)
Comment (GoOz) [inline src/styles/toolbar.styl]: Should be fixed with the Cozy-UI refactoring as $buttons doesn't exist anymore and the danger-outline is included in the $button placeholder… but until then I don't really have a clean solution :/

File: src/assets/icons/icon-trash-dark.svg (hunk 3 above)
Comment (GoOz) [inline src/assets/icons/icon-trash-dark.svg]: I'm too questioning that choice more and more lately. It's semantically more correct to use the background technique but not really the more pragmatic way. Any thought @m4dz?

File: src/styles/toolbar.styl (hunk 9 above)
Comment (y-lohse) [inline src/styles/toolbar.styl]: Alright, thanks. Good to know a fix is on the way :-)

File: src/actions/index.js (hunk 1 above)
Comment (goldoraf) [inline src/actions/index.js]: I understand, but I'd suggest to add a comment here, because the code is not very explicit.

File: src/assets/icons/icon-trash-dark.svg (hunk 3 above)
Comment (goldoraf) [inline src/assets/icons/icon-trash-dark.svg]: https://medium.com/@david.gilbertson/icons-as-react-components-de3e33cb8792#.xzmogaxo8
What do you think of this solution?

File: src/components/Nav.jsx (hunk 5 above)
Comment (goldoraf) [inline src/components/Nav.jsx]: Yeah, and the v4 of react-router should make this more easy to do.

File: src/containers/Trash.jsx [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+import { connect } from 'react-redux'
+import { withRouter } from 'react-router'
+
+import { openFolder, toggleFileSelection, showFileActionMenu } from '../actions'
+import { getVisibleFiles, mustShowSelectionBar, isBrowsingTrash } from '../reducers'
+import { TRASH_DIR_ID } from '../constants/config'
+
+import FileList from '../components/FileList'
+
+class Folder extends Component {
```
Comment (goldoraf) [inline src/containers/Trash.jsx]: It should be `class Trash` here ;)

**Comment by y-lohse** [comment]
Changes made, thanks for the review! Should be ready to merge now.
